### DISCUSSION TITLE
How do I use pyodide-build?

### DISCUSSION THREAD

**Discussion by wlach** [pyodide/pyodide#2007] [Q&A]
I want to experiment with creating a new pyodide package (pyrsistent), but I'm not sure how to access the [recommended way of doing so](https://pyodide.org/en/stable/usage/api/pyodide-build-cli.html) (`pyodide-build  mkpkg`). After successfully building pyodide, I do see a 
`./packages/.artifacts/bin/pyodide-build` got generated, but running it gives an error:

```
wlach@ece1d2689cfd:/src$ packages/.artifacts/bin/pyodide-build 
Traceback (most recent call last):
  File "/src/packages/.artifacts/bin/pyodide-build", line 33, in <module>
    sys.exit(load_entry_point('pyodide-build', 'console_scripts', 'pyodide-build')())
  File "/src/packages/.artifacts/bin/pyodide-build", line 22, in importlib_load_entry_point
    for entry_point in distribution(dist_name).entry_points
  File "/usr/local/lib/python3.9/importlib/metadata.py", line 524, in distribution
    return Distribution.from_name(distribution_name)
  File "/usr/local/lib/python3.9/importlib/metadata.py", line 187, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: pyodide-build
```

I tried pip installing it in a virtualenv, which gives another error:

```
wlach@oxherder:~/src/pyodide$ venv/bin/pip install pyodide-build
Collecting pyodide-build
  Downloading pyodide_build-0.18.1-py3-none-any.whl (26 kB)
Collecting cython<3.0
  Downloading Cython-0.29.24-cp39-cp39-manylinux1_x86_64.whl (1.9 MB)
     |████████████████████████████████| 1.9 MB 4.1 MB/s 
Collecting pyyaml
  Using cached PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (661 kB)
Installing collected packages: pyyaml, cython, pyodide-build
Successfully installed cython-0.29.24 pyodide-build-0.18.1 pyyaml-6.0
wlach@oxherder:~/src/pyodide$ venv/bin/pyodide-build 
Traceback (most recent call last):
  File "/home/wlach/src/pyodide/venv/bin/pyodide-build", line 8, in <module>
    sys.exit(main())
  File "/home/wlach/src/pyodide/venv/lib/python3.9/site-packages/pyodide_build/__main__.py", line 39, in main
    main_parser = make_parser()
  File "/home/wlach/src/pyodide/venv/lib/python3.9/site-packages/pyodide_build/__main__.py", line 33, in make_parser
    parser = module.make_parser(subparsers.add_parser(command_name))  # type: ignore
  File "/home/wlach/src/pyodide/venv/lib/python3.9/site-packages/pyodide_build/buildpkg.py", line 313, in make_parser
    default=common.get_make_flag("SIDE_MODULE_CFLAGS"),
  File "/home/wlach/src/pyodide/venv/lib/python3.9/site-packages/pyodide_build/common.py", line 44, in get_make_flag
    return get_make_environment_vars()[name]
KeyError: 'SIDE_MODULE_CFLAGS'
```

Can anyone give some guidance on what I should be doing here? I see https://github.com/pyodide/pyodide/issues/1977 but it seems to be more about what this could do in the future rather than what it does now. I seem to remember when I was working on pyodide a couple of years ago there was a `pyodide` cli tool which worked out of the pyodide checkout...

**Comment by wlach** [comment]
I half figured it out: it documents that you need to install this package locally in the [latest version of the build instructions](https://pyodide.org/en/latest/development/new-packages.html) (which, to be fair, the docs did suggest I do), although installing it locally doesn't work on my system (because I'm not root). I wound up installing it in a virtualenv (along with `ruamel.yaml`) and then managed to get it running.

**Reply by rth** [reply]
Hi,

Normally,
```py
pip install -e pyodide-build/
pyodide-build --help
```
in a virtualenv should work. It works for me on Linux.  We should indeed better indicate in the docs to always use a virtualenv.

It's strange you got the key error: what did you do to fix it?

**Reply by wlach** [reply]
> ```python
> pip install -e pyodide-build/
> pyodide-build --help
> ```
> 
> in a virtualenv should work. It works for me on Linux. We should indeed better indicate in the docs to always use a virtualenv.

:+1: or maybe set it up as part of the build?

> It's strange you got the key error: what did you do to fix it?

I didn't. :smile: I only got the local install working (inside a virtualenv).

**Reply by thadk** [reply]
Yes installing and building `pyodide-build` from the current 0.19-dev version (not 0.18.1) as above does seem to resolve the former #1724 and the meta.yaml `mkpkg` produces can be used with older versions even if they have broken `pyodide-build`'s. There is a slight difference in the new release's meta.yaml in that `PyPI` is used instead of `PyPi` as in older versions (0.18.1 and earlier) and that has to be adjusted.
